machine/v1/types_controlplanemachineset.go

```diff
@@ -284,6 +289,17 @@ type GCPFailureDomain struct {
 	Zone string `json:"zone"`
 }
 
+// OpenStackFailureDomain configures failure domain information for the OpenStack platform.
+type OpenStackFailureDomain struct {
```

Comment:
What if both `ComputeAvailabilityZone` and `StorageAvailabilityZone` are empty? What will be the default behavior for this case?

Reply:
* one or two fields set: valid
* both empty: valid, it means no failure domain. However in CCPMSO we check that if no failure domain was matched for a machine, the other masters can't match a failure domain. It's either we have 1+ failure domain or 0.

Reply:
@ingvagabund 

> What if both ComputeAvailabilityZone and StorageAvailabilityZone are empty?

The fields [are evaluated separately](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/195/files#diff-ae5bff491dfb3714be91613a81c7756db1ced45ca9698505b585b4a68ac346dbR38-R50).

If both fields are empty, then the Machine will not contain placement information, leaving Nova to act by its defaults.

Reply:
Note that as @EmilienM points out, CCPMSO will refuse to generate empty failureDomains. Those are however accepted if added by the user.

Reply:
I'll add validation to only accept a failure domain that has at least one of the fields to be set.